### PR TITLE
Support pysetiterator.

### PR DESCRIPTION
**Pull request by hatee-hatee-hatee-ho**
related https://github.com/RustPython/RustPython/issues/1652

I just fixed @dralley 's implementaion a little.

### CODE REVIEW

**Comment by hatee-hatee-hatee-ho** [comment]
while fix this issue, I encountered a test failure.
https://github.com/RustPython/RustPython/blob/1249fab930b99e505099656f1a62be518df23303/Lib/test/test_set.py#L1808-L1819

In CPython's implementation, `set.update` seems not using `__iter__`.
https://github.com/python/cpython/blob/da4e09fff6b483fe858997da5599c25397107ca1/Objects/setobject.c#L891-L894

**Comment by coolreader18** [comment]
I don't think that's exactly the problem; the code you linked to is specifically when the argument to update() is a dict, but a few lines below there's a call to `PyObject_GetIter()`. You could probably just mark that test as expectedFailure/skip for now; the fix for that is probably something in dictdatatype internals.

**Comment by coolreader18** [comment]
Hmm, it is strange though that it's only failing now.

**Comment by coolreader18** [comment]
Ohh, I see what you mean -- CPython skips iterating over the argument to update() when it's a set; it just merges them. Yeah, that would definitely be the fix to just add that optimization, if you'd like to.

File: vm/src/obj/objset.rs [hunk 1]
```diff
@@ -2,8 +2,9 @@
  * Builtin set type with a sequence of unique items.
  */
 use std::fmt;
+use std::sync::Arc;
```
Comment (youknowone) [inline vm/src/obj/objset.rs:5]:
```suggestion
use rustpython_common::rc::PyRc;
```

**Comment by skinny121** [comment]
That test case is for https://bugs.python.org/issue24583, which is a crash. So as long as it doesn't panic/hang, I think it's good.

File: vm/src/obj/objset.rs (hunk 1 above)
Comment (hatee-hatee-hatee-ho) [reply vm/src/obj/objset.rs]: Thanks, I fixed it.

File: vm/src/obj/objset.rs [hunk 2]
```diff
@@ -813,7 +818,68 @@ impl TryFromObject for SetIterable {
...
+impl fmt::Debug for PySetIterator {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        // TODO: implement more detailed, non-recursive Debug formatter
+        f.write_str("setiterator")
+    }
+}
+
+#[pyimpl]
+impl PySetIterator {
+    #[pymethod(name = "__next__")]
```
Comment (youknowone) [inline vm/src/obj/objset.rs:842]: `#[pymethod(magic)]` also works.
When funciton name is `f`, `magic` attribute turns the python attribute name into `__f__`.

File: vm/src/obj/objset.rs [hunk 3]
```diff
@@ -813,7 +818,68 @@ impl TryFromObject for SetIterable {
     }
 }
 
+#[derive(Copy, Clone, Default)]
+struct SetSizeInfo {
+    size: Option<usize>,
```
Comment (youknowone) [inline vm/src/obj/objset.rs:823]: it seems this is intended to be None when this is invalidated (changed) but not used yet, right?

Comment (hatee-hatee-hatee-ho) [reply vm/src/obj/objset.rs:823]: Oh, I forgot to invalidate it... I just fixed it.

File: vm/src/obj/objset.rs (hunk 2 above)
Comment (hatee-hatee-hatee-ho) [reply vm/src/obj/objset.rs]: Thanks, that looks great. I changed them to `#[pymethod(magic)]`.

**Comment by youknowone** [comment]
Thanks for contributing!

**Pull request merged**